Change blending job to MPMD to fix bugzilla 1593. Fix bugzilla 1226

YaliMao-NOAA opened the pull request:
Change blending job to MPMD to fix bugzilla 1593, meanwhile fix bugzilla 1226

The MPMD change for bugzilla 1593 is for NCO who wants to receive one single email
combining all forecast hours with missing UK data

For bugzilla 1226, AWC is fine with dbn_alert of US unblended data earlier in JWAFS_GRIB2_0P25 job

Bugzilla 1593 - Improve email notification for missing UK WAFS data
Bugzilla 1226 - Eliminate the duplicated dbn_alert for unblended gfs wafs data

Commit:
Change blending	job to MPMD to fix bugzilla 1593, meanwhile fix	bugzilla 1226

The MPMD change for bugzilla 1593 is for NCO who wants to receive one single email
combining all forecast hours with missing UK data

For bugzilla 1226, AWC is fine with dbn_alert of US unblended data earlier in JWAFS_GRIB2_0P25 job

Bugzilla 1593 -	Improve email notification for missing UK WAFS data
Bugzilla 1226 - Eliminate the duplicated dbn_alert for unblended gfs wafs data

Commit:
Update release note and modify the driver

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
-    else
-        sleep "${SLEEP_INT}"
-    fi
+fhours=${fhours:-"006 007 008 009 010 011 012 013 014 015 016 017 018 019 020 021 022 023 024 027 030 033 036 039 042 045 048"}
```
```suggestion
fhours="006 007 008 009 010 011 012 013 014 015 016 017 018 019 020 021 022 023 024 027 030 033 036 039 042 045 048"
```
`fhours` is set here.  There is no input for `fhours` from the calling script.

aerorahul replied:
You could generate this list with `seq` and make this an array.  Array has some nicer properties
```bash
fhours=($(seq -s ' ' -f "%03g" 6 1 24; seq -s ' ' -f "%03g" 27 3 48))

# get all elements
echo ${fhours[@]}  # 006 007 ... 048

# get number of elements 
echo ${#fhours[@]}  # 27
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
-        sleep "${SLEEP_INT}"
-    fi
+fhours=${fhours:-"006 007 008 009 010 011 012 013 014 015 016 017 018 019 020 021 022 023 024 027 030 033 036 039 042 045 048"}
+np=`echo $fhours | wc -w`
```
Do not use back-ticks "`"
```suggestion
np=$(echo $fhours | wc -w)
```

If `fhours` was an array, you would get this as:
```bash
np=${#fhours[@]}
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
-    fi
+fhours=${fhours:-"006 007 008 009 010 011 012 013 014 015 016 017 018 019 020 021 022 023 024 027 030 033 036 039 042 045 048"}
+np=`echo $fhours | wc -w`
+MPIRUN="mpiexec -np $np -cpu-bind verbose,core cfp"
```
```suggestion
MPIRUN="mpiexec -np ${np} -cpu-bind verbose,core cfp"
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
+
+rm -f wafsgrib2_0p25.cmdfile
+ic=0
+for fhr in $fhours ; do
```
Only running this on WCOSS2 where the `MP_PGMMODEL` is `mpmd`
```suggestion
export MP_PGMMODEL=mpmd
for fhr in ${fhours} ; do
```

If `fhours` was an array, it would be:
```bash
for fhr in ${fhours[@]}; do
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
+  else
+    echo "${USHwafs}/wafs_grib2_0p25_blending.sh $fhr > $DATA/${fhr}.log 2>&1">> wafsgrib2_0p25.cmdfile
+    export MP_PGMMODEL=mpmd
+  fi
```
This application is only going to run on wcoss2, where the scheduler is PBS and there is no `srun`.
Please remove the `if` block and retain the else.
Also, the MPMD filename should have `blending` since this is a blending script to be clear.
```suggestion
  echo "${USHwafs}/wafs_grib2_0p25_blending.sh ${fhr} > ${DATA}/${fhr}.log 2>&1" >> wafsgrib2_0p25_blending.cmdfile
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
+  fi
+  ic=$(expr $ic + 1)
 done
+$MPIRUN wafsgrib2_0p25.cmdfile
```
```suggestion
${MPIRUN} wafsgrib2_0p25_blending.cmdfile
export err=$?
if (( err != 0 )); then
  echo "FATAL ERROR: An error occured processing blending"
fi
for fhr in ${fhours}; do
  echo "=================== log file of fhr = ${fhr} ==================="
  cat "${DATA}/${fhr}.log"
done
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
+    cat "${COMOUT}/${RUN}.t${cyc}z.f${fhr}.wafs_blend_0p25_usonly.emailbody" | mail.py -s "${subject}" "${MAILTO}" -v
 fi
+
+for fhr in $fhours ; do
```
```suggestion
for fhr in ${fhours}; do
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
+
+for fhr in $fhours ; do
+    echo "=================== log file of fhr=$fhr ==================="
+    cat $DATA/${fhr}.log
```
```suggestion
    cat "${DATA}/${fhr}.log"
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
+    echo "${USHwafs}/wafs_grib2_0p25_blending.sh $fhr > $DATA/${fhr}.log 2>&1">> wafsgrib2_0p25.cmdfile
+    export MP_PGMMODEL=mpmd
+  fi
+  ic=$(expr $ic + 1)
```
What is the use of this `ic` here?

YaliMao-NOAA replied:
WAFS was once working on Hera. MPMD on Hera needs a sequential number (ic as an iterator) each line of cmdfile.

YaliMao-NOAA replied:
Oops, forget about it. I was thing about another loop.

aerorahul commented on ush/wafs_grib2_0p25_blending.sh:
```diff
@@ -0,0 +1,109 @@
+fhr=$1
+
+mkdir -p $DATA/$fhr
+cd "${DATA}/$fhr" || err_exit "FATAL ERROR: Could not 'cd ${DATA}/fhr'; ABORT!"
```
The error message was missing which fhr did the cd failed.
```suggestion
cd "${DATA}/${fhr}" || err_exit "FATAL ERROR: Could not 'cd ${DATA}/${fhr}'; ABORT!"
```

aerorahul commented on ush/wafs_grib2_0p25_blending.sh:
```diff
@@ -0,0 +1,109 @@
+    echo "turning back on dbn alert for unblended US WAFS product"
+elif [[ ! -f "${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2" ]]; then
+    echo "Warning: missing US unblended data - ${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2 "
+    exit # Silently quit if US data is missing
```
There shouldn't be a silent exit.

aerorahul commented on ush/wafs_grib2_0p25_blending.sh:
```diff
@@ -0,0 +1,109 @@
+if [[ "${SEND_UNBLENDED_US_WAFS}" == "YES" ]]; then
+    echo "turning back on dbn alert for unblended US WAFS product"
+elif [[ ! -f "${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2" ]]; then
+    echo "Warning: missing US unblended data - ${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2 "
```
```suggestion
    echo "WARNING: missing US unblended data - ${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2 "
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
+    cat "${COMOUT}/${RUN}.t${cyc}z.f${fhr}.wafs_blend_0p25_usonly.emailbody" | mail.py -s "${subject}" "${MAILTO}" -v
 fi
+
+for fhr in $fhours ; do
+    echo "=================== log file of fhr=$fhr ==================="
+    cat $DATA/${fhr}.log
+done
```
This section belongs right after the `MPIRUN`

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,69 @@
-        "${DBNROOT}/bin/dbn_alert" MODEL WAFS_0P25_BL_GB2 "${job}" "${COMOUT}/WAFS_0p25_blended_${PDY}${cyc}f${fhr}.grib2"
-    fi
+    cat mailmsg >"${COMOUT}/${RUN}.t${cyc}z.wafs_blend_0p25_usonly.emailbody"
+    cat "${COMOUT}/${RUN}.t${cyc}z.f${fhr}.wafs_blend_0p25_usonly.emailbody" | mail.py -s "${subject}" "${MAILTO}" -v
```
What is the value of `${fhr}` here?  This is not in a for loop over forecast hours.

Commit:
1. If US unblended data is missing, don't quit silently, instead send out email and dbn_alert.
2. Add not-blended email and dbn_alert if both UK and US unblended files are missing
3. Change fhours from a string to an array
4. Bug fix and code improvement

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,100 @@
-
-    # pick up US data
-    cpreq "${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2" .
+export err=$? ; err_chk
```
Doing `err_chk` here will abort the job.  This will result in no emails being sent.  Do not do `err_chk` here.

YaliMao-NOAA replied:
All missing data scenarios and exec errors are caught and handled in ush/wafs_grib2_0p25_blending.sh. If still some other error happens, there may be something abnormal and crashes the blending process, so no emails

aerorahul replied:
It is true they are handled there and this job should exit with a non-zero status.  But, there is mechanism in this job to alert people.  I believe that should happen regardless this job fails of whatever reason.

I would place the `err_chk` line at the end of the script.

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,100 @@
+rm -f wafsgrib2_0p25.cmdfile
+for fhr in ${fhours[@]}; do
+    echo "${USHwafs}/wafs_grib2_0p25_blending.sh $fhr > $DATA/${fhr}.log 2>&1">> wafsgrib2_0p25_blending.cmdfile
+    export MP_PGMMODEL=mpmd
```
exporting MP_PGMMODEL is independent of the forecast hour.  All it is doing inside the loop, is setting itself over and over.
```suggestion
export MP_PGMMODEL=mpmd
for fhr in ${fhours[@]}; do
    echo "${USHwafs}/wafs_grib2_0p25_blending.sh $fhr > $DATA/${fhr}.log 2>&1">> wafsgrib2_0p25_blending.cmdfile
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,100 @@
-    fi
+fhours=($(seq -s ' ' -f "%03g" 6 1 24; seq -s ' ' -f "%03g" 27 3 48))
+np=${#fhours[@]}
+MPIRUN="mpiexec -np ${np} -cpu-bind verbose,core cfp"
```
Declare this where it is actually used for clear readability
```suggestion
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,100 @@
-    . prep_step
+missing_uk_files="$(find $DATA -name 'missing_uk_files*')"
+missing_us_files="$(find $DATA -name 'missing_us_file*')"
+no_blending_files="$(find $DATA -name 'no_blending_files*')"
```
Why would it be `missing_uk_files` while it is `missing_us_file` when trying to `find`.  Be consistent to avoid confusion.

aerorahul commented on ush/wafs_grib2_0p25_blending.sh:
```diff
@@ -0,0 +1,145 @@
+    sleep 60
+    if [[ ! -f "${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2.idx" ]]; then
+	echo "WARNING: missing US unblended data - ${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2"
+	echo "Missing ${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2" > ../missing_us_file.$fhr
```
```suggestion
	echo "Missing ${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2" > ../missing_us_files.$fhr
```
Name the missing us file the same way the uk and no blending are; `files` instead of `file`

aerorahul commented on ush/wafs_grib2_0p25_blending.sh:
```diff
@@ -0,0 +1,145 @@
+
+fhr=$1
+
+mkdir -p $DATA/$fhr
```
```suggestion
mkdir -p "${DATA}/${fhr}"
```

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,100 @@
-    if ((err != 0)); then
-        echo "WARNING: WAFS blending 0p25 program failed at '${PDY}${cyc}f${fhr}'. Turning back on dbn alert for unblended US WAFS product"
-        SEND_UNBLENDED_US_WAFS="YES"
+if [[ ! -z "$missing_uk_files" ]] || [[ ! -z "$missing_us_files" ]] || [[ ! -z "$no_blending_files" ]] ; then
```
I don't think this logic is correct.
When there are no US and no UK, missing_uk_files and missing_us_files will be empty, and only no_blending files will be present.

On the other hand, if only UK is absent, this block will still send the same alert to AWC, even though US is present

YaliMao-NOAA replied:
If no_blending files, dbn_alert and email will be sent, causing different attentions of NCO. 
For example, at f007 if both UK and US are missing:
 1. At all other fhours, UK and US are present. Then only no_blending is alerted and emailed.
 2. UK data is absent at other fhours, both ukmissing and no_blending is alerted and emailed.

aerorahul reviewed (changes requested):
Please see detailed comments

Commit:
Update blending scripts

aerorahul commented on ush/wafs_grib2_0p25_blending.sh:
```diff
@@ -0,0 +1,145 @@
+    cat ../missing_uk_files.$fhr ../missing_us_files.$fhr > ../no_blending_files.$fhr
+    rm ../missing_uk_files.$fhr ../missing_us_files.$fhr
+elif [[ "${MISSING_UK_WAFS}" == "YES" ]]; then
+    echo "turning back on dbn alert for unblended US WAFS product"
+    # Avoid duplicate dbn_alert of unblended grib2 file which was done in the upstream grib2_0p25 job, fix bugzilla 1226
+    # "${DBNROOT}/bin/dbn_alert" MODEL WAFS_0P25_UBL_GB2 "${job}" "${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2"
+    "${DBNROOT}/bin/dbn_alert" MODEL WAFS_0P25_UBL_GB2_WIDX "${job}" "${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2.idx"
```
I suggest this section be right after line 63 where it has determined UK data is missing, and send the alert right there.
See similar comment for US below.

YaliMao-NOAA replied:
If line 98-101 is right after 63, we don't know whether the unblended US data is present yet

aerorahul replied:
a DBNAlert is not an abort.  The US data will be alerted when that is encountered.
If both are missing, then there are 2 alerts sent.
Moving the alert statements do not preclude from other not being present/absent

YaliMao-NOAA replied:
Will dbn_alert create an error message if the file to be alerted is not present?

aerorahul replied:
It is an alert, like an echo.

YaliMao-NOAA replied:
If both UK and US data are missing, moving these lines will make dbn_alert some file which doesn't exist.

aerorahul replied:
That is not how dbn_alert works.  Test with `fakedbn` to find out.

YaliMao-NOAA replied:
I just checked, dbn_alert needs 4 arguments:
/apps/ops/prod/nco/core/prod_util.v2.0.14/fakedbn/bin/dbn_alert
/apps/ops/prod/nco/core/prod_util.v2.0.14/para_dbn/bin/dbn_alert

aerorahul commented on ush/wafs_grib2_0p25_blending.sh:
```diff
@@ -0,0 +1,145 @@
+    # Three(3) unblended UK files for each cycle+fhour: icing, turb, cb
+    cat "${COMINuk}/egrr_wafshzds_unblended_"*"_0p25_${PDY:0:4}-${PDY:4:2}-${PDY:6:2}T${cyc}:00Z_t${fhr}.grib2" >"EGRR_WAFS_0p25_unblended_${PDY}_${cyc}z_t${fhr}.grib2"
+
+    if [[ "${MISSING_US_WAFS}" == "YES" ]]; then
+	if [ $SENDDBN = "YES" ] ; then
+	    "$DBNROOT/bin/dbn_alert" MODEL WAFS_UKMET_0P25_UBL_GB2 "${job}" "EGRR_WAFS_0p25_unblended_${PDY}_${cyc}z_t${fhr}.grib2"
+	fi
```
I suggest this section be right after line 87 where it has determined US data is missing, and send the alert right there.
See similar comment for UK above.

YaliMao-NOAA replied:
I can, but I think dbn_alert should be kept under the same block of code since the lines of missing UK can't be moved.

aerorahul replied:
This is US.  I don't follow

aerorahul commented on ush/wafs_grib2_0p25_blending.sh:
```diff
@@ -0,0 +1,145 @@
+fhr=$1
+
+mkdir -p "${DATA}/${fhr}"
+cd "${DATA}/${fhr}" || err_exit "FATAL ERROR: Could not 'cd ${DATA}/${fhr}'; ABORT!"
```
I think using `err_exit` is not appropriate here, because if a single forecast hour has a problem, the whole job will be aborted.
`err_exit` is handled in the exscript right after the `MPIRUN` call
```suggestion
cd "${DATA}/${fhr}" || err=1
if (( err != 0 )); then
  echo "FATAL ERROR: Could not 'cd ${DATA}/${fhr}'"
  echo "WAFS blending 0p25 program failed at " ${PDY}${cyc}F${fhr} > ../no_blending_files.$fhr
  exit 1
fi
```
This will allow other hours to continue.

aerorahul commented on ush/wafs_grib2_0p25_blending.sh:
```diff
@@ -0,0 +1,145 @@
+	if (( err != 0 )); then
+	    echo "turning back on dbn alert for unblended US WAFS product"
+	    "${DBNROOT}/bin/dbn_alert" MODEL WAFS_0P25_UBL_GB2_WIDX "${job}" "${COMINus}/WAFS_0p25_unblended_${PDY}${cyc}f${fhr}.grib2.idx"
+	    echo "WAFS blending 0p25 program failed at " ${PDY}${cyc}F${ffhr} > ../no_blending_files.$fhr
```
There is no variable named `ffhr`
```suggestion
	    echo "WAFS blending 0p25 program failed at ${PDY}${cyc}F${fhr}" > "../no_blending_files.${fhr}"
```

YaliMao-NOAA replied:
Thank you!

aerorahul reviewed (changes requested):
some comments and caught a bug

Commit:
Bug fix

aerorahul commented on ush/wafs_grib2_0p25_blending.sh:
```diff
@@ -0,0 +1,145 @@
+fhr=$1
+
+mkdir -p "${DATA}/${fhr}"
+cd "${DATA}/${fhr}" || err=1
```
What does this do?  Setting `err=1` will be useless and the script is going to fail.  This is not correct.

YaliMao-NOAA replied:
Sorry I thought it was what you suggested. I must have missed something

aerorahul replied:
I think my suggestions got dismissed when you made the commit.

YaliMao-NOAA replied:
I missed other lines. I am going to add them.

Commit:
Update the way of handling err and removing np variable for MPIRUN

aerorahul commented on scripts/exwafs_grib2_0p25_blending.sh:
```diff
@@ -2,152 +2,104 @@
+    cat "${COMOUT}/${RUN}.t${cyc}z.wafs_blend_0p25_noblending.emailbody" | mail.py -s "${subject}" "${MAILTO}" -v
 fi
+
+export err=$? ; err_chk
```
If you reach here, `err` will almost be guaranteed to be 0.
```suggestion
err_chk
```

YaliMao-NOAA committed:
Update scripts/exwafs_grib2_0p25_blending.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

aerorahul reviewed (approved):
looks good. thanks for including comments and suggestions.
